dev/sdk.go

```diff
@@ -92,3 +96,99 @@ func (dev *DaggerDev) introspection(ctx context.Context, installer func(*dagger.
 		WithExec([]string{"codegen", "introspect", "-o", "/schema.json"}).
 		File("/schema.json"), nil
 }
+
+type gitPublishOpts struct {
+	sdk string
+
+	source, dest       string
+	sourceTag, destTag string
+	sourcePath         string
+	sourceFilter       string
+	sourceEnv          *dagger.Container
+
+	username    string
+	email       string
+	githubToken *dagger.Secret
+
+	dryRun bool
+}
+
+func gitPublish(ctx context.Context, opts gitPublishOpts) error {
+	base := opts.sourceEnv
+	if base == nil {
+		base = dag.Container().
+			From(consts.AlpineImage).
+			WithExec([]string{"apk", "add", "-U", "--no-cache", "git", "go", "python3"})
+	}
+
+	// FIXME: move this into std modules
+	git := base.
+		WithExec([]string{"git", "config", "--global", "user.name", opts.username}).
+		WithExec([]string{"git", "config", "--global", "user.email", opts.email})
+	if !opts.dryRun {
+		githubTokenRaw, err := opts.githubToken.Plaintext(ctx)
+		if err != nil {
+			return err
+		}
+		encodedPAT := base64.URLEncoding.EncodeToString([]byte("pat:" + githubTokenRaw))
+		git = git.
+			WithEnvVariable("GIT_CONFIG_COUNT", "1").
+			WithEnvVariable("GIT_CONFIG_KEY_0", "http.https://github.com/.extraheader").
+			WithSecretVariable("GIT_CONFIG_VALUE_0", dag.SetSecret("GITHUB_HEADER", fmt.Sprintf("AUTHORIZATION: Basic %s", encodedPAT)))
+	}
+
+	result := git.
+		WithEnvVariable("CACHEBUSTER", identity.NewID()).
+		WithWorkdir("/src/dagger").
+		WithExec([]string{"git", "clone", opts.source, "."}).
+		WithExec([]string{"git", "fetch", "origin", "-v", "--update-head-ok", fmt.Sprintf("refs/*%[1]s:refs/*%[1]s", strings.TrimPrefix(opts.sourceTag, "refs/"))}).
+		WithEnvVariable("FILTER_BRANCH_SQUELCH_WARNING", "1").
+		WithExec([]string{
+			"git", "filter-branch", "-f", "--prune-empty",
+			"--subdirectory-filter", opts.sourcePath,
+			"--tree-filter", opts.sourceFilter,
+			"--", opts.sourceTag,
+		})
+	if !opts.dryRun {
+		result = result.WithExec([]string{
+			"git",
+			"push",
+			// "--force", // NOTE: disabled to avoid accidentally rewriting the history
```

Comment:
Should we remove `--force` entirely?

Reply:
Going to leave it in - technically, if `main` or a tag is force-pushed, then we need the `--force` flag here. However, we shouldn't really do this in general.